File: ruby/05/KataFive.rb
require 'digest/sha2'

class BloomFilter
  def initialize(filename=nil, step=3)
    @step = step
    @bit_array = Array.new(2 ** (8 * step), false)
    File.foreach(filename) { |word| insert(word.strip) } unless filename.nil?
  end

  def hashes(string)
    hash_val = Digest::SHA512.digest(string.downcase)
    pattern = "H#{@step * 2}" * (hash_val.size / @step)
    hash_val.unpack(pattern).map { |x| x.hex }
  end

  def insert(string)
    hashes(string).each { |hash_val| @bit_array[hash_val] = true }
  end

  def contains(string)
    hashes(string).inject(true) { |acc, hash_val| acc && @bit_array[hash_val] }
  end
end

File: python/05/KataFive.py
import hashlib
import binascii


class BloomFilter(object):
    def __init__(self, filename=None):
        self.step = 3
        self.bit_array = [False] * pow(2, (8 * self.step))
        if filename:
            with open(filename) as f:
                for word in f:
                    self.insert(word.strip())

    @staticmethod
    def bytes2int(bytes):
        return int(binascii.hexlify(bytes), 16)

    def hashes(self, string):
        hash_val = hashlib.sha512(string.encode()).digest()
        return [self.bytes2int(hash_val[i:i + self.step]) for i in range(0, len(hash_val) - 1, self.step)]

    def insert(self, string):
        for hash_val in self.hashes(string.lower()):
            self.bit_array[hash_val] = True

    def contains(self, string):
        for hash_val in self.hashes(string.lower()):
            if not self.bit_array[hash_val]:
                return False
        return True